features/mbedtls/targets/TARGET_Samsung/sha/sha256_alt.c

```diff
@@ -0,0 +1,428 @@
+#define SHA256_VALIDATE(cond)  MBEDTLS_INTERNAL_VALIDATE( cond )
+
+/*
+ * 32-bit integer manipulation macros (big endian)
+ */
+#ifndef GET_UINT32_BE
+#define GET_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (n) = ( (uint32_t) (b)[(i)    ] << 24 )             \
+        | ( (uint32_t) (b)[(i) + 1] << 16 )             \
+        | ( (uint32_t) (b)[(i) + 2] <<  8 )             \
+        | ( (uint32_t) (b)[(i) + 3]       );            \
+} while( 0 )
+#endif
+
+#ifndef PUT_UINT32_BE
+#define PUT_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
+    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
+    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
+    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
+} while( 0 )
+#endif
+
+
+#include "string.h"
+#include "sss_common.h"
+#include "mb_cmd_system.h"
+
+int mbedtls_sha256_sw_finish_ret(mbedtls_sha256_context *ctx, unsigned char output[32]);
+int mbedtls_sha256_sw_update_ret(mbedtls_sha256_context *ctx, const unsigned char *input, size_t ilen);
+
+void mbedtls_sha256_init(mbedtls_sha256_context *ctx)
+{
+	memset( ctx, 0, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_free(mbedtls_sha256_context *ctx)
+{
+    if (ctx == NULL) {
+        return;
+    }
+	memset( ctx, 0, sizeof( mbedtls_sha256_context ) );
+}
+
+void mbedtls_sha256_clone(mbedtls_sha256_context *dst,
+                          const mbedtls_sha256_context *src)
+{
+    // Corner case: Destination/source contexts are the same
+    if (dst == src) {
+        return;
+    }
+
+	//SHA256_VALIDATE( dst != NULL );
+	//SHA256_VALIDATE( src != NULL );
+
+	memcpy(dst, src, sizeof(mbedtls_sha256_context));
+}
+
+/*
+ * SHA-256 context setup
+ */
+int mbedtls_sha256_starts_ret(mbedtls_sha256_context *ctx, int is224)
+{
+    SHA256_VALIDATE_RET( ctx != NULL );
+    SHA256_VALIDATE_RET( is224 == 0 || is224 == 1 );
+
+    ctx->total[0] = 0;
+    ctx->total[1] = 0;
+
+    if( is224 == 0 )
+    {  //SHA-256 by SSS H/W
+		ctx->is224 = 0;
+	///	memset( ctx->sbuf, 0, sizeof( ctx->sbuf ) );
+		memset( ctx, 0, sizeof( mbedtls_sha256_context ) );
+	}
+    else
+    {
+        /* SHA-224 */
+        ctx->state[0] = 0xC1059ED8;
+        ctx->state[1] = 0x367CD507;
+        ctx->state[2] = 0x3070DD17;
+        ctx->state[3] = 0xF70E5939;
+        ctx->state[4] = 0xFFC00B31;
+        ctx->state[5] = 0x68581511;
+        ctx->state[6] = 0x64F98FA7;
+        ctx->state[7] = 0xBEFA4FA4;
+		ctx->is224 = is224;
+    }
+
+    return( 0 );
+}
+
+/*
+ * SHA-256 process buffer
+ */
+int mbedtls_sha256_update_ret(mbedtls_sha256_context *ctx, const unsigned char *input, size_t ilen)
+{
+	if(ctx->is224){
+		mbedtls_sha256_sw_update_ret(ctx, input, ilen);
+	}
+	else {
+		if(ilen > MAX_MB_HASH_BLOCK_BLEN || ctx->totals > MAX_MB_HASH_BLOCK_BLEN) {
```

Comment:
If `ilen > MAX_MB_HASH_BLOCK_BLEN` is true, then it will enter line 129 even if `ctx->totals == 0`. So from my perspective, it may not be logical conflict, however, I request @jh6186 for additional code review.

Reply:
That code will be used for continuous update case, in case first ilen(e.g., 128) is less then BLOCK_BLEN (256) and next update with ilen(129) makes ctx->totals (257) exceeds BLOCK_BLEN , then S/W fallback would follows. This is what i intended. Please leave that condition function, it will improve but not in this moment.